framework/config/src/main/java/org/apache/cloudstack/framework/config/ConfigKeyScheduledExecutionWrapper.java

```diff
@@ -0,0 +1,111 @@
+
+import static java.util.concurrent.TimeUnit.NANOSECONDS;
+
+/**
+ * Uses a ScheduledExecutorService and config key to execute a runnable,
+ * dynamically rescheduling based on the long value of the config key.
+ * Timing is similar to ScheduledExecutorService.scheduleAtFixedRate(),
+ * but we look up the next runtime dynamically via the config key.
+ * <p>
+ * If config key is zero, this disables the execution. We skip execution
+ * and check once a minute in order to re-start execution if re-enabled.
+ */
+public class ConfigKeyScheduledExecutionWrapper implements Runnable {
+    protected Logger logger = LogManager.getLogger(getClass());
+    private final ScheduledExecutorService executorService;
+    private final Runnable command;
+    private final ConfigKey<?> configKey;
+    private final TimeUnit unit;
+    private long enableIntervalSeconds = 60;
+
+    private void validateArgs(ScheduledExecutorService executorService, Runnable command, ConfigKey<?> configKey) {
+        if (executorService == null) {
+            throw new IllegalArgumentException("ExecutorService cannot be null");
+        }
+        if (command == null) {
+            throw new IllegalArgumentException("Command cannot be null");
+        }
+        if (configKey == null) {
+            throw new IllegalArgumentException("ConfigKey cannot be null");
+        }
+    }
+
+    public ConfigKeyScheduledExecutionWrapper(ScheduledExecutorService executorService, Runnable command,
+            ConfigKey<?> configKey, TimeUnit unit) {
+        validateArgs(executorService, command, configKey);
+        this.executorService = executorService;
+        this.command = command;
+        this.configKey = configKey;
+        this.unit = unit;
+    }
+
+    protected ConfigKeyScheduledExecutionWrapper(ScheduledExecutorService executorService, Runnable command,
+            ConfigKey<?> configKey, int enableIntervalSeconds, TimeUnit unit) {
+        validateArgs(executorService, command, configKey);
+        this.executorService = executorService;
+        this.command = command;
+        this.configKey = configKey;
+        this.unit = unit;
+        this.enableIntervalSeconds = enableIntervalSeconds;
+    }
+
+    public ScheduledFuture<?> start() {
+        long duration = getConfigValue();
+        duration = duration < 0 ? 0 : duration;
+        return this.executorService.schedule(this, duration, this.unit);
+    }
+
+    long getConfigValue() {
+        if (this.configKey.value() instanceof Long) {
+            return (Long) this.configKey.value();
+        } else if (this.configKey.value() instanceof Integer) {
+            return (Integer) this.configKey.value();
+        } else {
+            throw new IllegalArgumentException("ConfigKey value must be a Long or Integer");
+        }
+    }
+
+    @Override
+    public void run() {
+        if (getConfigValue() <= 0) {
+            executorService.schedule(this, enableIntervalSeconds, TimeUnit.SECONDS);
```

Comment:
what happens if `ResourceCountCheckInterval` was set to a negative value before starting the management server? Should we defend against that?

Reply:
If the value is negative, we use `enableIntervalSeconds` as the duration for schedule.

Reply:
Ok, I was confusing the two constructors, I see the protected one at line 69 is getting a `enableIntervalSeconds` but doesn't check its value. I think a guard against negative values there would help.

On line 228 from `ResourceLimitManagerImpl` the constructor with the default `enableIntervalSeconds` is used. Can we remove the other constructor? I don't see any use of it.